R/install-remote.R

```diff
@@ -58,3 +81,126 @@ is.remote <- function(x) inherits(x, "remote")
+    remote_sha <- remote_sha(remote)
+  }
+
+  if (is.null(local_sha)) {
+    local_sha <- local_sha(remote_package_name(remote))
+  }
+
+  same <- remote_sha == local_sha
+  same <- isTRUE(same) && !is.na(same)
+  !same
+}
+
+local_sha <- function(name) {
+  if (!is_installed(name)) {
+    return(NA_character_)
+  }
+  package2remote(name)$sha %||% NA_character_
+}
+
+package2remote <- function(name, lib = .libPaths(), repos = getOption("repos"), type = getOption("pkgType")) {
+
+  x <- tryCatch(utils::packageDescription(name, lib.loc = lib), error = function(e) NA, warning = function(e) NA)
+
+  # will be NA if not installed
+  if (identical(x, NA)) {
+    return(remote("cran",
+        name = name,
+        repos = repos,
+        pkg_type = type,
+        sha = NA_character_))
+  }
+
+  if (is.null(x$RemoteType)) {
+
+    # Packages installed with install.packages() or locally without devtools
+    return(remote("cran",
+        name = x$Package,
+        repos = repos,
+        pkg_type = type,
+        sha = x$Version))
+  }
+
+  switch(x$RemoteType,
+    github = remote("github",
+      host = x$RemoteHost,
+      repo = x$RemoteRepo,
+      subdir = x$RemoteSubdir,
+      username = x$RemoteUsername,
+      ref = x$RemoteRef,
+      sha = x$RemoteSha),
+    gitlab = remote("gitlab",
+      host = x$RemoteHost,
+      repo = x$RemoteRepo,
+      subdir = x$RemoteSubdir,
+      username = x$RemoteUsername,
+      ref = x$RemoteRef,
+      sha = x$RemoteSha),
+    git = remote("git",
+      url = x$RemoteUrl,
+      ref = x$RemoteRef,
+      sha = x$RemoteSha,
+      subdir = x$RemoteSubdir),
+    bitbucket = remote("bitbucket",
+      host = x$RemoteHost,
+      repo = x$RemoteRepo,
+      username = x$RemoteUsername,
+      ref = x$RemoteRef,
+      sha = x$RemoteSha,
+      subdir = x$RemoteSubdir),
+    svn = remote("svn",
+      url = x$RemoteUrl,
+      svn_subdir = x$RemoteSubdir,
+      revision = x$RemoteSha,
+      args = x$RemoteArgs),
+    local = remote("local",
+      path = x$RemoteUrl,
+      subdir = x$RemoteSubdir,
+      sha = {
+        # Packages installed locally might have RemoteSha == NA_character_
+        x$RemoteSha %||% x$Version
+      }),
+    url = remote("url",
+      url = x$RemoteUrl,
+      subdir = x$RemoteSubdir,
+      config = x$RemoteConfig,
+      pkg_type = x$RemotePkgType),
+    bioc = remote("bioc",
+      repo = x$RemoteRepo,
+      mirror = x$RemoteMirror,
+      release = x$RemoteRelease,
+      username = x$RemoteUsername,
+      password = x$RemotePassword,
+      revision = x$RemoteRevision,
+      sha = x$RemoteSha),
+
+    # packages installed with install_cran
+    cran = remote("cran",
+      name = x$Package,
+      repos = eval(parse(text = x$RemoteRepos)),
+      pkg_type = x$RemotePkgType,
+      sha = x$RemoteSha))
+}
+
+#' @export
+format.remotes <- function(x, ...) {
+  vapply(x, format, character(1))
+}
+
+#' @export
+`[.remotes` <- function(x,i,...) {
```

Comment:
Why is this trick needed? Can you  please add  it  as  internal docs, or  a comment?

Reply:
It is possible it is not needed, I will remove it and see if anything breaks.